torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -198,6 +202,102 @@ bool is_numpy_scalar(PyObject* obj) {
+  std::vector<int64_t> sizes;
+  ScalarType dtype;
+  npy_intp data_ptr;
+  std::vector<int64_t> strides;
+  try {
+    if (!PyDict_Check(cuda_dict)) {
+      throw TypeError("`__cuda_array_interface__` must be a dict");
+    }
+    PyObject *py_shape = PyDict_GetItemString(cuda_dict, "shape");
+    PyObject *py_typestr = PyDict_GetItemString(cuda_dict, "typestr");
+    PyObject *py_data = PyDict_GetItemString(cuda_dict, "data");
+    if (py_shape == nullptr || py_typestr == nullptr || py_data == nullptr) {
+      throw TypeError("attributes `shape`, `typestr`, and `data` must exist");
+    }
+
+    // Extract the `shape` attribute
+    int ndim = PySequence_Length(py_shape);
+    if (ndim == -1) {
+      throw TypeError("shape must be a sequence");
+    }
+    std::vector<npy_intp> npy_shape(ndim);
+    PyArray_IntpFromSequence(py_shape, npy_shape.data(), ndim);
+    sizes = to_aten_shape(ndim, npy_shape.data());
+
+    // Extract the `typestr` attribute
+    int dtype_size_in_bytes;
+    {
+      PyArray_Descr *descr;
+      if(!PyArray_DescrConverter(py_typestr, &descr)) {
+        throw ValueError("cannot parse `typestr`");
+      }
+      dtype = numpy_dtype_to_aten(descr->type_num);
+      dtype_size_in_bytes = descr->elsize;
+      assert(dtype_size_in_bytes > 0);
+    }
+
+    // Extract the `data` attribute
+    if(!PyTuple_Check(py_data) || PyTuple_GET_SIZE(py_data) != 2) {
+      throw TypeError("`data` must be a 2-tuple of (int, bool)");
+    }
+    PyErr_Clear();
+    data_ptr = PyArray_PyIntAsIntp(PyTuple_GET_ITEM(py_data, 0));
+    npy_bool read_only;
+    if (PyErr_Occurred() || !PyArray_BoolConverter(PyTuple_GET_ITEM(py_data, 1), &read_only)) {
+      throw TypeError("`data` must be a 2-tuple (int, bool)");
+    }
+    if (read_only) {
+      throw TypeError("the read only flag is not supported");
+    }
+
+    // Extract the `strides` attribute
+    PyObject *py_strides = PyDict_GetItemString(cuda_dict, "strides");
+    if (py_strides != nullptr && py_strides != Py_None) {
+      if (PySequence_Length(py_shape) == -1 || PySequence_Length(py_shape) != ndim) {
+        throw TypeError("strides must be a sequence of the same length as shape");
+      }
+      std::vector<npy_intp> npy_strides(ndim);
+      PyArray_IntpFromSequence(py_strides, npy_strides.data(), ndim);
+      strides = to_aten_shape(ndim, npy_strides.data());
+    } else {
+      strides = at::detail::defaultStrides(sizes);
+    }
+
+    // __cuda_array_interface__ strides use bytes. Torch strides use element counts.
+    for (auto& stride : strides) {
+      if (stride%dtype_size_in_bytes != 0) {
+        throw ValueError(
+            "given array strides not a multiple of the element byte size. "
+            "Make a copy of the array to reallocate the memory.");
+        }
+      stride /= dtype_size_in_bytes;
+    }
+  } catch (const ValueError &e) {
+    Py_DECREF(cuda_dict);
```

Comment:
This "looks" as if it frees `cuda_dict` on error conditions, but actually because you are only catching `ValueError`, all of the `TypeError` exit points leak memory. To make matters worse, you never free `cuda_dict` on the success case; even more leaky memory!

Reply:
Fixed in 2cf1117